### PR TITLE
[BACKLOG-7324, BACKLOG-7407] New Karaf port and data cache conflict r…

### PR DESCRIPTION
**Pull request by pentaho-nbaker**
…esolution strategies

 Port conflicts and managed with a new ServerSocket based approach

 Karaf caches are now identified by PDI client type to prevent re-use of the wrong type

### CODE REVIEW

**Comment by lucboudreau** [comment]
http://10.100.11.222:8080/job/wingman/26/

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. ant -Dtestreports.dir=bin/reports/unit-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco && ant -Dpentaho.coding.standards.version=1.0.4 -Dtestreports.dir=bin/reports/integration-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove



Changed files:
==============
  extensions/src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/osgi/KarafBoot.java
  extensions/src/org/pentaho/platform/osgi/KarafInstance.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePort.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePortFactory.java
  extensions/src/org/pentaho/platform/osgi/KarafInstanceResolverException.java
  extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/settings/PortAssigner.java
  extensions/src/org/pentaho/platform/settings/ServerPort.java
  extensions/src/org/pentaho/platform/settings/ServerPortDto.java
  extensions/src/org/pentaho/platform/settings/ServerPortRegistry.java
  extensions/test-res/KarafInstanceTest/KarafPorts.yaml
  extensions/test-res/karafBootTest/system/karaf/etc/custom.properties
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedHelper.java
  extensions/test-src/org/pentaho/platform/osgi/KarafBootTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstancePortFactoryTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstanceTest.java
  extensions/test-src/org/pentaho/platform/settings/PortAssignerTest.java
  extensions/test-src/org/pentaho/platform/settings/ServerPortRegistryTest.java

Newly Broken Tests:
======================
org.pentaho.platform.settings.PortAssignerTest.initializationError:
-----------------------------------------------
java.lang.Exception
```
java.lang.Exception: No runnable methods
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)

```


Still Broken Tests:
======================
org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider:
-----------------------------------------------
junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
Expected: iterable containing ["org.apache", "org.apache.log4j"]
     but: item 0: was "org.apache.log4j"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider(SystemPackageExtrapolatorTest.java:85)

```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException(ArchiveLoaderTest.java:93)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory(ArchiveLoaderTest.java:66)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```

Unit test coverage change:
============
org.pentaho.platform.osgi.ExceptionBasedClientTypeProvider:
 * Branch New Coverage : 80.00%
 * Complexity New Coverage : 75.00%
 * Instruction New Coverage : 91.25%
 * Line New Coverage : 94.44%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.KarafBoot:
 * Branch Change :  + .55%
 * Complexity Change :  + 1.24%
 * Instruction Change :  + .37%
 * Line Change : **-.72%**

org.pentaho.platform.osgi.KarafInstance:
 * Branch Change : **-12.09%**
 * Complexity Change :  + 6.36%
 * Instruction Change : **-7.45%**
 * Line Change :  + 2.29%
 * Method Change :  + 17.22%

org.pentaho.platform.osgi.KarafInstance.new FilenameFilter() {...}:
 * Branch Change : **-60.00%**
 * Complexity Change : **-42.86%**
 * Instruction Change : **-92.86%**
 * Line Change : **-92.31%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.KarafInstancePort:
 * Instruction Change : **-3.62%**
 * Line Change :  + 2.73%

org.pentaho.platform.osgi.KarafInstancePortFactory:
 * Branch Change : **-100.00%**
 * Complexity Change : **-100.00%**
 * Instruction Change : **-75.00%**
 * Line Change : **-86.67%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver:
 * Branch New Coverage : 55.88%
 * Complexity New Coverage : 52.17%
 * Instruction New Coverage : 79.44%
 * Line New Coverage : 84.62%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver.new FilenameFilter() {...}:
 * Complexity New Coverage : 100.00%
 * Instruction New Coverage : 100.00%
 * Line New Coverage : 100.00%
 * Method New Coverage : 100.00%

org.pentaho.platform.settings.PortAssigner:
 * Branch Change : **-58.70%**
 * Complexity Change : **-62.16%**
 * Instruction Change : **-82.97%**
 * Line Change : **-86.02%**
 * Method Change : **-100.00%**

org.pentaho.platform.settings.ServerPort:
 * Complexity Change :  + 6.67%
 * Instruction Change :  + 15.32%
 * Line Change :  + 9.03%
 * Method Change : **-1.82%**

org.pentaho.platform.settings.ServerPortRegistry:
 * Complexity Change :  + 14.29%
 * Instruction Change :  + 9.38%
 * Line Change :  + 6.45%
 * Method Change :  + 20.00%

License header violations:
==========================
1. extensions/src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProvider.java
2. extensions/src/org/pentaho/platform/osgi/IClientTypeProvider.java
3. extensions/src/org/pentaho/platform/osgi/IKarafInstanceResolver.java
4. extensions/src/org/pentaho/platform/osgi/KarafInstanceResolverException.java
5. extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java
6. extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java
7. extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedHelper.java
8. extensions/test-src/org/pentaho/platform/settings/ServerPortRegistryTest.java

Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards
File /home/buildguy/project/extensions/src/org/pentaho/platform/osgi/KarafBoot.java

1. Message: 'switch' is not followed by whitespace. at Line: 325 Column: 13

File /home/buildguy/project/extensions/src/org/pentaho/platform/settings/ServerPort.java

1. Message: Unused import - org.slf4j.Logger. at Line: 20 Column: 8
2. Message: Unused import - org.slf4j.LoggerFactory. at Line: 21 Column: 8

File /home/buildguy/project/extensions/src/org/pentaho/platform/settings/ServerPortDto.java

1. Message: Unused import - java.util.ArrayList. at Line: 21 Column: 8
2. Message: Unused import - java.util.List. at Line: 22 Column: 8
3. Message: '(' is not followed by whitespace. at Line: 33 Column: 17
4. Message: '=' is not preceded with whitespace. at Line: 33 Column: 21
5. Message: '=' is not followed by whitespace. at Line: 33 Column: 22
6. Message: ')' is not preceded with whitespace. at Line: 33 Column: 34
7. Message: '(' is not followed by whitespace. at Line: 34 Column: 18
8. Message: ')' is not preceded with whitespace. at Line: 34 Column: 37

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java

1. Message: File does not end with a newline. at Line: 0
2. Message: '{' is not preceded with whitespace. at Line: 34 Column: 25

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedHelper.java

1. Message: ')' is not preceded with whitespace. at Line: 7 Column: 64
2. Message: '{' is not preceded with whitespace. at Line: 7 Column: 65

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/osgi/KarafInstancePortFactoryTest.java

1. Message: Unused import - org.mockito.Mockito.mock. at Line: 23 Column: 15
2. Message: Unused import - org.mockito.Mockito.times. at Line: 24 Column: 15
3. Message: Unused import - org.mockito.Mockito.verify. at Line: 25 Column: 15
4. Message: Unused import - java.io.File. at Line: 27 Column: 8
5. Message: Unused import - java.io.FileInputStream. at Line: 28 Column: 8
6. Message: Unused import - java.io.FileNotFoundException. at Line: 29 Column: 8
7. Message: Unused import - java.io.InputStream. at Line: 31 Column: 8
8. Message: Unused import - java.util.List. at Line: 32 Column: 8
9. Message: Unused import - java.util.Map. at Line: 33 Column: 8
10. Message: Unused import - org.yaml.snakeyaml.Yaml. at Line: 41 Column: 8

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/osgi/KarafInstanceTest.java

1. Message: Unused import - org.mockito.invocation.InvocationOnMock. at Line: 23 Column: 8
2. Message: Unused import - org.mockito.stubbing.Answer. at Line: 24 Column: 8
3. Message: Unused import - org.pentaho.platform.settings.PortAssigner. at Line: 25 Column: 8
4. Message: Unused import - org.mockito.Mockito.doAnswer. at Line: 32 Column: 15
5. Message: Unused import - org.mockito.Mockito.mock. at Line: 33 Column: 15
6. Message: Unused import - org.mockito.Mockito.when. at Line: 34 Column: 15

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/settings/PortAssignerTest.java

1. Message: Unused import - org.junit.Assert.assertTrue. at Line: 19 Column: 15
2. Message: Unused import - java.io.File. at Line: 21 Column: 8
3. Message: Unused import - org.junit.Test. at Line: 22 Column: 8
4. Message: Unused import - edu.emory.mathcs.backport.java.util.Arrays. at Line: 24 Column: 8

File: extensions/src/org/pentaho/platform/settings/PortAssigner.java [hunk 1]
```diff
@@ -245,4 +216,8 @@ public void clear() {
     usedPorts.clear();
     usedPortsExternal.clear();
   }
+
+  public static Integer getFreePortStartingAt( Integer startPort ) {
```
Comment (rmansoor) [inline extensions/src/org/pentaho/platform/settings/PortAssigner.java]: @pentaho-nbaker why do we need this method now ?

File: extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        continue;
+      }
+      int instanceNo = Integer.parseInt( matcher.group( 1 ) );
+
+      maxInstanceNoFound = Math.max( maxInstanceNoFound, instanceNo );
+
+      File lockFile = new File( dataDirectory, ".lock" );
+      FileLock lock = null;
+      if ( !lockFile.exists() ) {
+        locked = false;
```
Comment (tkafalas) [inline extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java]: @pentaho-nbaker Maybe use DATA static here

File: extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+    String cacheParentFolder = instance.getCacheParentFolder();
+    File clientTypeCacheFolder = new File( cacheParentFolder + "/" + instance.getClientType() );
+    clientTypeCacheFolder.mkdirs();
+    File[] dataDirectories = clientTypeCacheFolder.listFiles( new FilenameFilter() {
+      @Override public boolean accept( File dir, String name ) {
+        return name.startsWith( DATA );
+      }
+    } );
+    int maxInstanceNoFound = 0;
+    Pattern pattern = Pattern.compile( "data\\-([0-9]+)" );
```
Comment (tkafalas) [inline extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java]: @pentaho-nbaker  Maybe use the DATA static here

File: extensions/src/org/pentaho/platform/settings/PortAssigner.java (hunk 1 above)
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/settings/PortAssigner.java]: The whole class can go

File: extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java (hunk 3 above)
Comment (pentaho-nbaker) [inline extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java]: Definitely. Thanks I'll do that

**Comment by lucboudreau** [comment]
http://10.100.11.222:8080/job/wingman/35/

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. ant -Dtestreports.dir=bin/reports/unit-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco && ant -Dpentaho.coding.standards.version=1.0.4 -Dtestreports.dir=bin/reports/integration-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove



Changed files:
==============
  extensions/src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/osgi/KarafBoot.java
  extensions/src/org/pentaho/platform/osgi/KarafInstance.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePort.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePortFactory.java
  extensions/src/org/pentaho/platform/osgi/KarafInstanceResolverException.java
  extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/settings/PortAssigner.java
  extensions/src/org/pentaho/platform/settings/ServerPort.java
  extensions/src/org/pentaho/platform/settings/ServerPortDto.java
  extensions/src/org/pentaho/platform/settings/ServerPortRegistry.java
  extensions/test-res/KarafInstanceTest/KarafPorts.yaml
  extensions/test-res/karafBootTest/system/karaf/etc/custom.properties
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedHelper.java
  extensions/test-src/org/pentaho/platform/osgi/KarafBootTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstancePortFactoryTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstanceTest.java
  extensions/test-src/org/pentaho/platform/settings/PortAssignerTest.java
  extensions/test-src/org/pentaho/platform/settings/ServerPortRegistryTest.java

Newly Broken Tests:
======================
org.pentaho.platform.settings.PortAssignerTest.initializationError:
-----------------------------------------------
java.lang.Exception
```
java.lang.Exception: No runnable methods
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)

```


Still Broken Tests:
======================
org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider:
-----------------------------------------------
junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
Expected: iterable containing ["org.apache", "org.apache.log4j"]
     but: item 0: was "org.apache.log4j"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider(SystemPackageExtrapolatorTest.java:85)

```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException(ArchiveLoaderTest.java:93)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory(ArchiveLoaderTest.java:66)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```

Unit test coverage change:
============
org.pentaho.platform.osgi.ExceptionBasedClientTypeProvider:
 * Branch New Coverage : 80.00%
 * Complexity New Coverage : 75.00%
 * Instruction New Coverage : 91.25%
 * Line New Coverage : 94.44%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.KarafBoot:
 * Branch Change :  + .56%
 * Complexity Change :  + 1.23%
 * Instruction Change :  + .37%
 * Line Change : **-.71%**

org.pentaho.platform.osgi.KarafInstance:
 * Branch Change : **-12.09%**
 * Complexity Change :  + 6.36%
 * Instruction Change : **-7.45%**
 * Line Change :  + 2.29%
 * Method Change :  + 17.22%

org.pentaho.platform.osgi.KarafInstance.new FilenameFilter() {...}:
 * Branch Change : **-60.00%**
 * Complexity Change : **-42.86%**
 * Instruction Change : **-92.86%**
 * Line Change : **-92.31%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.KarafInstancePort:
 * Instruction Change : **-3.62%**
 * Line Change :  + 2.73%

org.pentaho.platform.osgi.KarafInstancePortFactory:
 * Branch Change : **-100.00%**
 * Complexity Change : **-100.00%**
 * Instruction Change : **-75.00%**
 * Line Change : **-86.67%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver:
 * Branch New Coverage : 55.88%
 * Complexity New Coverage : 52.17%
 * Instruction New Coverage : 79.44%
 * Line New Coverage : 84.62%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver.new FilenameFilter() {...}:
 * Complexity New Coverage : 100.00%
 * Instruction New Coverage : 100.00%
 * Line New Coverage : 100.00%
 * Method New Coverage : 100.00%

org.pentaho.platform.settings.PortAssigner:
 * Branch Change : **-58.70%**
 * Complexity Change : **-62.16%**
 * Instruction Change : **-82.97%**
 * Line Change : **-86.02%**
 * Method Change : **-100.00%**

org.pentaho.platform.settings.ServerPort:
 * Complexity Change :  + 6.67%
 * Instruction Change :  + 15.32%
 * Line Change :  + 9.03%
 * Method Change : **-1.82%**

org.pentaho.platform.settings.ServerPortRegistry:
 * Complexity Change :  + 14.29%
 * Instruction Change :  + 9.38%
 * Line Change :  + 6.45%
 * Method Change :  + 20.00%

License header violations:
==========================
1. extensions/src/org/pentaho/platform/osgi/KarafInstanceResolverException.java
2. extensions/test-src/org/pentaho/platform/settings/ServerPortRegistryTest.java

Checkstyle violations:
======================
See: https://github.com/pentaho/pentaho-coding-standards
File /home/buildguy/project/extensions/src/org/pentaho/platform/osgi/KarafBoot.java

1. Message: 'switch' is not followed by whitespace. at Line: 328 Column: 13

File /home/buildguy/project/extensions/src/org/pentaho/platform/settings/ServerPort.java

1. Message: Unused import - org.slf4j.Logger. at Line: 20 Column: 8
2. Message: Unused import - org.slf4j.LoggerFactory. at Line: 21 Column: 8

File /home/buildguy/project/extensions/src/org/pentaho/platform/settings/ServerPortDto.java

1. Message: Unused import - java.util.ArrayList. at Line: 21 Column: 8
2. Message: Unused import - java.util.List. at Line: 22 Column: 8
3. Message: '(' is not followed by whitespace. at Line: 33 Column: 17
4. Message: '=' is not preceded with whitespace. at Line: 33 Column: 21
5. Message: '=' is not followed by whitespace. at Line: 33 Column: 22
6. Message: ')' is not preceded with whitespace. at Line: 33 Column: 34
7. Message: '(' is not followed by whitespace. at Line: 34 Column: 18
8. Message: ')' is not preceded with whitespace. at Line: 34 Column: 37

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java

1. Message: File does not end with a newline. at Line: 0

File /home/buildguy/project/extensions/test-src/org/pentaho/platform/settings/PortAssignerTest.java

1. Message: Unused import - org.junit.Assert.assertTrue. at Line: 19 Column: 15
2. Message: Unused import - java.io.File. at Line: 21 Column: 8
3. Message: Unused import - org.junit.Test. at Line: 22 Column: 8
4. Message: Unused import - edu.emory.mathcs.backport.java.util.Arrays. at Line: 24 Column: 8

**Comment by lucboudreau** [comment]
http://10.100.11.222:8080/job/wingman/36/

**Comment by lucboudreau** [comment]
Build Commands:
==============
Build:
------
1. ant -Dtestreports.dir=bin/reports/unit-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml clean-all resolve jacoco && ant -Dpentaho.coding.standards.version=1.0.4 -Dtestreports.dir=bin/reports/integration-test -Djunit.jvmargs=-XX:PermSize=256m -f extensions/build.xml jacoco-integration checkstyle publish-local

Cleanup:
-------
rm -r ~/.ivy2/local || echo no publish local to remove



Changed files:
==============
  extensions/src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IClientTypeProvider.java
  extensions/src/org/pentaho/platform/osgi/IKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/osgi/KarafBoot.java
  extensions/src/org/pentaho/platform/osgi/KarafInstance.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePort.java
  extensions/src/org/pentaho/platform/osgi/KarafInstancePortFactory.java
  extensions/src/org/pentaho/platform/osgi/KarafInstanceResolverException.java
  extensions/src/org/pentaho/platform/osgi/OSGIBoot.java
  extensions/src/org/pentaho/platform/osgi/ServerSocketBasedKarafInstanceResolver.java
  extensions/src/org/pentaho/platform/osgi/SystemPackageExtrapolator.java
  extensions/src/org/pentaho/platform/settings/PortAssigner.java
  extensions/src/org/pentaho/platform/settings/ServerPort.java
  extensions/src/org/pentaho/platform/settings/ServerPortDto.java
  extensions/src/org/pentaho/platform/settings/ServerPortRegistry.java
  extensions/test-res/KarafInstanceTest/KarafPorts.yaml
  extensions/test-res/karafBootTest/system/karaf/etc/custom.properties
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedClientTypeProviderTest.java
  extensions/test-src/org/pentaho/platform/osgi/ExceptionBasedHelper.java
  extensions/test-src/org/pentaho/platform/osgi/KarafBootTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstancePortFactoryTest.java
  extensions/test-src/org/pentaho/platform/osgi/KarafInstanceTest.java
  extensions/test-src/org/pentaho/platform/settings/PortAssignerTest.java
  extensions/test-src/org/pentaho/platform/settings/ServerPortRegistryTest.java

Still Broken Tests:
======================
org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider:
-----------------------------------------------
junit.framework.AssertionFailedError
```
junit.framework.AssertionFailedError: 
Expected: iterable containing ["org.apache", "org.apache.log4j"]
     but: item 0: was "org.apache.log4j"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at org.pentaho.platform.osgi.SystemPackageExtrapolatorTest.testJBossProvider(SystemPackageExtrapolatorTest.java:85)

```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillContinueToLoadOnException(ArchiveLoaderTest.java:93)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```
org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory:
-----------------------------------------------
org.mockito.exceptions.verification.junit.ArgumentsAreDifferent
```
Argument(s) are different! Wanted:
file.renameTo(
    
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoaderTest.testWillImportAllZipsInADirectory(ArchiveLoaderTest.java:66)
Actual invocation has different arguments:
file.renameTo(
    /root/path/jobs.zip.197001021017
);
-> at org.pentaho.platform.plugin.services.importer.ArchiveLoader.loadAll(ArchiveLoader.java:68)

	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.pentaho.platform.plugin.services.import
```

Unit test coverage change:
============
org.pentaho.platform.osgi.ExceptionBasedClientTypeProvider:
 * Branch New Coverage : 80.00%
 * Complexity New Coverage : 75.00%
 * Instruction New Coverage : 91.25%
 * Line New Coverage : 94.44%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.KarafBoot:
 * Branch Change :  + .56%
 * Complexity Change :  + 1.23%
 * Instruction Change :  + .37%
 * Line Change : **-.71%**

org.pentaho.platform.osgi.KarafInstance:
 * Branch Change : **-12.09%**
 * Complexity Change :  + 6.36%
 * Instruction Change : **-7.45%**
 * Line Change :  + 2.29%
 * Method Change :  + 17.22%

org.pentaho.platform.osgi.KarafInstance.new FilenameFilter() {...}:
 * Branch Change : **-60.00%**
 * Complexity Change : **-42.86%**
 * Instruction Change : **-92.86%**
 * Line Change : **-92.31%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.KarafInstancePort:
 * Instruction Change : **-3.62%**
 * Line Change :  + 2.73%

org.pentaho.platform.osgi.KarafInstancePortFactory:
 * Branch Change : **-100.00%**
 * Complexity Change : **-100.00%**
 * Instruction Change : **-75.00%**
 * Line Change : **-86.67%**
 * Method Change : **-100.00%**

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver:
 * Branch New Coverage : 55.88%
 * Complexity New Coverage : 52.17%
 * Instruction New Coverage : 79.44%
 * Line New Coverage : 84.62%
 * Method New Coverage : 100.00%

org.pentaho.platform.osgi.ServerSocketBasedKarafInstanceResolver.new FilenameFilter() {...}:
 * Complexity New Coverage : 100.00%
 * Instruction New Coverage : 100.00%
 * Line New Coverage : 100.00%
 * Method New Coverage : 100.00%

org.pentaho.platform.plugin.services.metadata.PentahoMetadataDomainRepository:
 * Branch Change :  + 1.25%
 * Complexity Change :  + .80%
 * Instruction Change :  + .93%
 * Line Change :  + .82%

org.pentaho.platform.settings.PortAssigner:
 * Branch Change : **-58.70%**
 * Complexity Change : **-62.16%**
 * Instruction Change : **-82.97%**
 * Line Change : **-86.02%**
 * Method Change : **-100.00%**

org.pentaho.platform.settings.ServerPort:
 * Complexity Change :  + 6.67%
 * Instruction Change :  + 15.32%
 * Line Change :  + 9.03%
 * Method Change : **-1.82%**

org.pentaho.platform.settings.ServerPortRegistry:
 * Complexity Change :  + 14.29%
 * Instruction Change :  + 9.38%
 * Line Change :  + 6.45%
 * Method Change :  + 20.00%

License header violations:
==========================
1. extensions/src/org/pentaho/platform/osgi/SystemPackageExtrapolator.java

**Pull request merged**